redis/redis
# Enabled background and reply time tracking on blocked on keys/blocked on background work clients

This PR enables tracking time of the background tasks and on replies, opening the door for properly tracking commands that rely on blocking/background work via the slowlog, latency history, and commandstats. 

Some notes:
- The time spent blocked waiting for key changes, or blocked on synchronous replication is not accounted for. 

- **This PR does not affect latency tracking of commands that are non-blocking or do not have background work.** ( meaning that it all stays the same with exception to `BZPOPMIN`,`BZPOPMAX`,`BRPOP`,`BLPOP`, etc... and module's commands that rely on background threads ). 

-  Specifically for latency history command we've added a new event class named `command-unblocking` that will enable latency monitoring on commands that spawn background threads to do the work.

- For blocking commands we're now considering the total time of a command as the time spent on call() + the time spent on replying when unblocked.

- For Modules commands that rely on background threads we're now considering the total time of a command as the time spent on call (main thread) + the time spent on the background thread ( if marked within `RedisModule_MeasureTimeStart()` and `RedisModule_MeasureTimeEnd()` ) + the time spent on replying (main thread)


## test
To test for this feature we've added a `unit/moduleapi/blockonbackground` test that relies on a module that blocks the client and sleeps on the background for a given time. 
- check blocked command that uses RedisModule_MeasureTimeStart() is tracking background time
- check blocked command that uses RedisModule_MeasureTimeStart() is tracking background time even in timeout
- check blocked command with multiple calls RedisModule_MeasureTimeStart()  is tracking the total background time
- check blocked command without calling RedisModule_MeasureTimeStart() is not reporting background time

## Samples
### current behaviour ( not tracking time spent on background ):
```
127.0.0.1:6379> config set slowlog-log-slower-than 100000
OK
127.0.0.1:6379> slowlog len
(integer) 0
127.0.0.1:6379> block.debug 0 10000
(integer) 343176910
127.0.0.1:6379> slowlog len
(integer) 0
127.0.0.1:6379> block.debug 2000 10000
(integer) 1207730440
(2.00s)
127.0.0.1:6379> slowlog len
(integer) 0
```
#### commandstats example
```
127.0.0.1:6379> block.debug 2000 10000
(integer) 1768881473
(2.00s)
127.0.0.1:6379> block.debug 4000 10000
(integer) 630830182
(4.00s)
127.0.0.1:6379> info commandstats
# Commandstats
cmdstat_block.debug:calls=2,usec=226,usec_per_call=113.00
```

### new behaviour ( tracking time spend on background ):

```
127.0.0.1:6379> config set slowlog-log-slower-than 100000
OK
127.0.0.1:6379> slowlog len
(integer) 0
127.0.0.1:6379> block.debug 0 10000
(integer) 1471490433
127.0.0.1:6379> slowlog len
(integer) 0
127.0.0.1:6379> block.debug 2000 10000
(integer) 1127835047
(2.00s)
127.0.0.1:6379> slowlog len
(integer) 1
127.0.0.1:6379> slowlog get 1
1) 1) (integer) 0
   2) (integer) 1593712320
   3) (integer) 2053204
   4) 1) "block.debug"
      2) "2000"
      3) "10000"
   5) "127.0.0.1:51818"
   6) ""
```
#### commandstats example
```
127.0.0.1:6379> block.debug 2000 10000
(integer) 1052512224
(2.00s)
127.0.0.1:6379> block.debug 4000 10000
(integer) 215961991
(4.00s)
127.0.0.1:6379> info commandstats
# Commandstats
cmdstat_block.debug:calls=2,usec=6099877,usec_per_call=3049938.50
```
#### latency monitor example
```
127.0.0.1:6379> config set latency-monitor-threshold 200
OK
127.0.0.1:6379> block.debug 500 1000
(integer) 1887208743
(0.50s)
127.0.0.1:6379> latency history command-unblocking
1) 1) (integer) 1611265472
   2) (integer) 500
```
